shared/src/main/scala/frankenpaxos/epaxos/Replica.scala

```diff
@@ -0,0 +1,558 @@
+  }
+
+  private def checkSameDependencies(preAcceptResponses: mutable.ListBuffer[PreAcceptOk], numSameNeeded: Int, total: Int): Tuple2[Boolean, Option[Seq[Instance]]] = {
+    val dependenciesNumCount: mutable.Map[Seq[Instance], Int] = mutable.Map()
+    for (preacceptOk <- preAcceptResponses) {
+      dependenciesNumCount.put(preacceptOk.dependencies, dependenciesNumCount.getOrElse(preacceptOk.dependencies, 0) + 1)
+    }
+
+    for (key <- dependenciesNumCount.keys) {
+      if (dependenciesNumCount.getOrElse(key, 0) >= numSameNeeded) {
+        return Tuple2[Boolean, Option[Seq[Instance]]](true, Some(key))
+      }
+    }
+
+    Tuple2[Boolean, Option[Seq[Instance]]](false, None)
+  }
+
+  private def handlePreAcceptOk(address: Transport#Address, value: PreAcceptOk): Unit = {
+    // Verify leader is receiving responses
+    if (value.commandInstance.leaderIndex == index) {
+      //preacceptResponses.put(value.commandInstance, preacceptResponses.getOrElse(value.commandInstance, 0) + 1)
+      val preAcceptOkSet: mutable.ListBuffer[PreAcceptOk] = preacceptResponses.getOrElse(value.commandInstance, mutable.ListBuffer.empty)
+      preAcceptOkSet.append(value)
+      preacceptResponses.put(value.commandInstance, preAcceptOkSet)
+      val quorumSize: Int = (config.replicaAddresses.size / 2) + 1
+      //println(value + " " + preacceptResponses)
+      if (preacceptResponses.getOrElse(value.commandInstance, mutable.ListBuffer.empty).size >= quorumSize) {
+        //println("Reached quorum size")
+        val N: Int = config.replicaAddresses.size
+        val seqTuple: Tuple2[Boolean, Option[Int]] = checkSameSequenceNumbers(
+          preacceptResponses.getOrElse(value.commandInstance, mutable.ListBuffer.empty), N - 2, N)
+        val depsTuple: Tuple2[Boolean, Option[Seq[Instance]]] = checkSameDependencies(
+          preacceptResponses.getOrElse(value.commandInstance, mutable.ListBuffer.empty), N - 2, N)
+        /*val sameSequenceNumber: Boolean = checkSameSequenceNumbers(preacceptResponses.getOrElse(value.commandInstance, mutable.Set.empty), N - 2, N)
+        val sameDeps: Boolean = checkSameDependencies(preacceptResponses.getOrElse(value.commandInstance, mutable.Set.empty), N - 2, N)*/
+        if (seqTuple._1 && depsTuple._1 && !value.avoid) {
+          // Send request reply to client
+          val clientAddress: Option[Transport#Address] = instanceClientMapping.get(value.commandInstance)
+          if (clientAddress.nonEmpty) {
+            val client = chan[Client[Transport]](clientAddress.get, Client.serializer)
+            client.send(ClientInbound().withRequestReply(RequestReply(
+              value.command,
+              value.commandInstance
+            )))
+          }
+          if (value.dependencies.size > 0) {
+           println(value.dependencies)
+          }
+          // Run commit phase
+          runCommitPhase(value.command, seqTuple._2.get, depsTuple._2.get, value.commandInstance)
+        } else {
```

Comment:
If I'm understanding this logic correctly, as soon as a replica receives a quorum of votes, it will always fail to have a superquorum of matching votes and immediately enter the Paxos-Accept phase? I think you should wait for a quorum and then set a timer for fast quorums. If you get a fast quorum before the timer goes off, great! Cancel the timer and run `runCommitPhase`. Otherwise, move on to the `startPaxosAcceptPhase`.